Source/Core/Core/IOS/USB/Emulated/Skylander.cpp

```diff
@@ -0,0 +1,721 @@
+        // indexed would be 0x20.
+
+        // 4th byte is the block to write to.
+
+        // Bytes 5 - 20 ({ 01, 02, 03, 04, 05, 06, 07, 08, 09, 0a, 0b, 0c, 0d, 0e, 0f }) are the
+        // data to write.
+
+        // The response does not appear to return the id of the Skylander being written, the 2nd
+        // byte is 0x00; however, the 3rd byte echos the block that was written (0x00 in example
+        // above.)
+
+      case 'W':
+      {
+        const u8 sky_num = buf[1] & 0xF;
+        const u8 block = buf[2];
+        g_skyportal.WriteBlock(sky_num, block, &buf[3], q_result.data());
+        q_queries.push(q_result);
+        q_data = {buf[0],  buf[1],  buf[2],  buf[3],  buf[4],  buf[5],  buf[6],
+                  buf[7],  buf[8],  buf[9],  buf[10], buf[11], buf[12], buf[13],
+                  buf[14], buf[15], buf[16], buf[17], buf[18]};
+        cmd->expected_count = 19;
+        break;
+      }
+      default:
+        ERROR_LOG_FMT(IOS_USB, "Unhandled Skylander Portal Query: {}", buf[0]);
+        break;
+      }
+      break;
+    case 0x0A:
+      cmd->expected_count = 8;
+      break;
+    case 0x0B:
+      cmd->expected_count = 8;
+      break;
+    default:
+      ERROR_LOG_FMT(IOS_USB, "Unhandled Request {}", cmd->request);
+      break;
+    }
+    break;
+
+  default:
+    break;
+  }
+  cmd->expected_time = Common::Timer::NowUs() + 100;
+  GetTransferThread().AddTransfer(std::move(cmd), q_data);
+  return 0;
+}
+int SkylanderUSB::SubmitTransfer(std::unique_ptr<BulkMessage> cmd)
+{
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}] Bulk: length={} endpoint={}", m_vid, m_pid,
+                m_active_interface, cmd->length, cmd->endpoint);
+  return 0;
+}
+
+// When an Interrupt Message is received by the Skylander Portal from the console,
+// it needs to respond with either the status of the console if there are no Control Messages that
+// require a response, or with the relevant response data that is requested from the Control
+// Message.
+int SkylanderUSB::SubmitTransfer(std::unique_ptr<IntrMessage> cmd)
+{
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}] Interrupt: length={} endpoint={}", m_vid, m_pid,
+                m_active_interface, cmd->length, cmd->endpoint);
+
+  auto& system = Core::System::GetInstance();
+  auto& memory = system.GetMemory();
+  u8* buf = memory.GetPointerForRange(cmd->data_address, cmd->length);
+  std::array<u8, 64> q_result = {};
+  // Audio requests are 64 bytes long, are the only Interrupt requests longer than 32 bytes,
+  // echo the request as the response and respond after 1ms
+  if (cmd->length > 32)
+  {
+    std::array<u8, 64> q_audio_result = {};
+    u8* audio_buf = q_audio_result.data();
+    memcpy(audio_buf, buf, cmd->length);
+    cmd->expected_time = Common::Timer::NowUs() + 1000;
+    cmd->expected_count = cmd->length;
+    GetTransferThread().AddTransfer(std::move(cmd), q_audio_result);
```

Comment:
Surprised I didn’t notice this until now. Since audio data is time sensitive and needs to be handled as quickly as possible, no response is sent back to the portal. No need for a transfer.

Reply:
Yeah took me a while to figure this one out too - the RPCS3 code didn't handle audio and instead replied saying that the audio version doesn't exist, but managed to figure it out with my physical Trap Team portal!

Reply:
So sending this should be removed right?

Reply:
Oh no this still needs to be sent back - if nothing is sent back then the Wii isn't aware if the audio has been played. It just needs to echo the request as a response to the Wii, and then the Wii sends a new request with new audio data, which is what I have done in the code

Reply:
Weird… I’ll have to look into that. In my research I came to the conclusion that no response is needed.

Reply:
I guess it isn't really a response, but if you don't echo the request back the game stalls out. Not sure if that is a Dolphin specific issue or maybe a Wii thing, but every request requires a response, usually just acknowledgement that the request has been received to continue sending requests, which is what happens for all of the control transfers anyway

Reply:
Interesting. I’ll look into it.